runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/middleware/FlexibleChecksumsResponseMiddleware.kt

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.http.middleware
+
+import aws.smithy.kotlin.runtime.ClientException
+import aws.smithy.kotlin.runtime.hashing.toHashFunction
+import aws.smithy.kotlin.runtime.http.HttpBody
+import aws.smithy.kotlin.runtime.http.isSuccess
+import aws.smithy.kotlin.runtime.http.operation.*
+import aws.smithy.kotlin.runtime.http.response.HttpCall
+import aws.smithy.kotlin.runtime.http.toCompletingBody
+import aws.smithy.kotlin.runtime.http.toHashingBody
+import aws.smithy.kotlin.runtime.io.*
+import aws.smithy.kotlin.runtime.util.*
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.Deferred
+import kotlin.coroutines.coroutineContext
+
+// The priority to validate response checksums, if multiple are present
+internal val CHECKSUM_HEADER_VALIDATION_PRIORITY_LIST: List<String> = listOf(
+    "x-amz-checksum-crc32c",
+    "x-amz-checksum-crc32",
+    "x-amz-checksum-sha1",
+    "x-amz-checksum-sha256",
+)
+
+internal val MULTIPART_CHECKSUM_HEADER_REGEX = Regex("x-amz-checksum-[a-zA-Z0-9]+-[0-9]+")
+
+/**
+ * Validate a flexible checksums response.
+ *
+ * Wraps the response in a hashing body, calculating the checksum as the response is streamed to the user.
+ * The checksum is validated after the user has consumed the entire body.
+ * Users can see which checksum was validated by referencing the `ResponseChecksumValidated` execution context variable.
+ */
+@InternalApi
+public class FlexibleChecksumsResponseMiddleware : ReceiveMiddleware {
+
+    public companion object {
+        public val ResponseChecksum: AttributeKey<Deferred<String>> = AttributeKey("ResponseChecksum")
```

Comment:
**fix**: documentation for these keys if they are public and expected to be used